### PR TITLE
add functions to model cache for instance mutater apiserver functionaltiy 

### PR DESCRIPTION
**Pull request by hmlanigan**
Based on PR9966 for change to cache.Machine constructer.

## Description of change

Added charm.LXDProfile(), unit.Application(), machine.Units(), and application.CharmURL() to model cache.

## QA steps

No change  to current juju usage.

### CODE REVIEW

File: core/cache/watcher.go [hunk 1]
```diff
@@ -146,6 +148,188 @@ func (w *ConfigWatcher) configChanged(topic string, value interface{}) {
...
+	if len(values) != 2 {
+		logger.Errorf("programming error, 2 values not provided")
+		return
+	}
+	appName, chURL := values[0], values[1]
+	info, ok := w.applications[appName]
+	if ok {
+		ch, err := w.getCharm(chURL)
+		if err != nil {
+
```
Comment (SimonRichardson) [inline core/cache/watcher.go]: What should happen here? I guess return with the error?

File: core/cache/watcher.go [hunk 2]
```diff
@@ -146,6 +148,188 @@ func (w *ConfigWatcher) configChanged(topic string, value interface{}) {
...
+func (w *MachineAppLXDProfileWatcher) addUnit(unit *Unit) bool {
+	_, ok := w.applications[unit.details.Application]
+	if !ok {
+		info := appInfo{
+			charmURL: unit.details.CharmURL,
+			units:    set.NewStrings(unit.details.Name),
+		}
+		ch, err := w.getCharm(unit.details.CharmURL)
+		if err != nil {
+
```
Comment (SimonRichardson) [inline core/cache/watcher.go]: Missing handling error?

File: core/cache/watcher.go [hunk 3]
```diff
@@ -146,6 +148,188 @@ func (w *ConfigWatcher) configChanged(topic string, value interface{}) {
...
+	} else {
+		w.applications[unit.details.Application].units.Add(unit.details.Name)
+	}
+	if w.applications[unit.details.Application].charmProfile != nil {
+		return true
+	}
+	return false
+}
+
+func (w *MachineAppLXDProfileWatcher) removeUnit(names []string) bool {
```
Comment (SimonRichardson) [inline core/cache/watcher.go]: I'm not sure I like this, I think we should either do `[2]string`, `name struct{UnitName, AppName string}` or pass two arguments.

File: core/cache/watcher.go [hunk 4]
```diff
@@ -146,6 +148,188 @@ func (w *ConfigWatcher) configChanged(topic string, value interface{}) {
...
+			w.notify()
+		}
+	}(&notify)
+
+	names, okString := value.([]string)
+	unit, okUnit := value.(*Unit)
+	switch {
+	case okString:
+		logger.Tracef("Stop watching %q", names)
+		notify = w.removeUnit(names)
```
Comment (SimonRichardson) [inline core/cache/watcher.go]: I think this should be... and validate the logic here, as removeUnit becomes very cryptic further down in the PR.

```
notify = w.removeUnit(names[0], names[1])
```

File: core/cache/watcher.go (hunk 1 above)
Comment (hmlanigan) [reply core/cache/watcher.go]: resolved in pr 9966 lxdprofilewatcher.go

File: core/cache/watcher.go (hunk 2 above)
Comment (hmlanigan) [reply core/cache/watcher.go]: will be resolved in pr 9966 lxdprofilewatcher.go

File: core/cache/watcher.go (hunk 3 above)
Comment (hmlanigan) [reply core/cache/watcher.go]: passing two args, will be resolved in pr 9966 lxdprofilewatcher.go

File: core/cache/watcher.go (hunk 4 above)
Comment (hmlanigan) [reply core/cache/watcher.go]: will be resolved in pr 9966 lxdprofilewatcher.go

**Comment by hmlanigan** [comment]
squashed two commits before:
$$merge$$

**Pull request merged**